DLPX-86536 CIS: sshd configuration

justsanjeev opened the pull request:
[Uploading Compliance_Report_scan_report_cis_10_110_247_248_dephx3sr1_20240112.pdf…]()
<!--
<details open>
<summary><h2> Background </h2></summary>

Provide a clear description of the high-level effort with which this
pull request is associated. Recall that while anyone in the organization
can see this pull request, not everyone necessarily has the same context
as you. If applicable, link to design documents or high-level tracking
epics here.
</details>
-->

<details open>
<summary><h2> Problem </h2></summary>

`sshd` configurations in Delphix Engine is not CIS complaint.
</details>

<!--
<details>
<summary><h2> Evaluation </h2></summary>

If the cause of the problem is not obvious, provide a root-cause
analysis (RCA), ideally using the Five Whys iterative interrogative
technique or some other form of analytical reasoning.
</details>
-->

<details open>
<summary><h2> Solution </h2></summary>
Set the below values in the `sshd` configuration 

- `MaxAuthTries` :  6
- `Ciphers`: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
- `KexAlgorithms`: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256
- `ClientAliveInterval`: 180 (Azure), 300 (AWS) 
- `LoginGraceTime`: 120
- `MaxStartups`: 10:30:60
- `PermitRootLogin`: "no"


- `sshd_config` is set to have a mode 600

We have earlier removed `PermitRootLogin` and the CIS Scan was failed so we have now everything in place and the scan results are also passed for all the below sections 

    
 
</details>


<details>
<summary><h2> Testing Done </h2></summary>
Performed a few tests as below, used an images created from `git ab-pre-push` and from the job 


#### http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/7747/


Used the image from above jenkins run 

````
(awscli) sanjeev.rohila@Sanjeev-Rohilas-MacBook-Pro QA-39385-Test-Data % dc groups list | grep dlpx-sanjeev.rohila-develop
dlpx-sanjeev.rohila-develop                      2024-01-11-19-33  Delphix Engine 20.0.0.0  5621 1ecb3442
(awscli) sanjeev.rohila@Sanjeev-Rohilas-MacBook-Pro QA-39385-Test-Data % 
````

 to create an appliance and performed below tests -

````
delphix@ip-10-110-247-248:~$ sudo sshd -T | grep -i ClientAliv
clientaliveinterval 300
clientalivecountmax 3
delphix@ip-10-110-247-248:~$ 
````
As per report (CIS) attached all the below sections are passed 

````
(5.25) 2158 Status of the 'Permissions' settings for the 'sshd_config' file - PASSED
(5.34) 2234 Status of the 'MaxAuthTries' setting in the 'sshd_config' file - PASSED
(5.37) 2239 Status of the 'PermitRootLogin' setting in the 'sshd_config' file - PASSED
(5.40) 5220 Status of the 'Ciphers' setting in the 'sshd_config' file - PASSED
(5.41) 5283 Status of the 'MACs' setting in the '/etc/ssh/sshd_config' file - PASSED
(5.42) 14400 Status of the 'kexalgorithms' setting within the /etc/ssh/sshd_config file - PASSED
(5.44) 5222 Status of the 'ClientAliveInterval' setting in the 'sshd_config' file - PASSED
(5.45) 5281 Status of the 'LoginGraceTime' setting in the '/etc/ssh/sshd_config' file - PASSED
(5.49) 5284 Status of the 'MaxStartups' setting in the '/etc/ssh/sshd_config' file - PASSED
````

</details>

<!--
<details>
<summary><h2> Implementation </h2></summary>

Describe the implementation details of the solution. Generally the
reasoning behind any non-obvious code should be recorded in code
comments. However, code comments should always describe the current
state of the code; in contrast, this section may be useful to describe
the relationship between the original code and the code being proposed
in this pull request.
</details>
-->

<!--
<details>
<summary><h2> Notes to Reviewers </h2></summary>

Provide any extra information a reviewer may need to know before
evaluating your pull request. For example, you may wish to describe
which files should be reviewed first or which files are auto-generated.
If the review tool cannot detect a file move, you may wish to link to a
patch comparing the old file and the new file.
</details>
-->

<!--
<details>
<summary><h2> Deployment Plan </h2></summary>

Describe how the code in this pull request will be deployed. Some
deployments are complicated and may require flag days between multiple
repositories or the provisioning of new infrastructure. Describing your
deployment plan allows reviewers to ensure that your work will not cause
any unnecessary interruption to end users.
</details>
-->

<!--
<details>
<summary><h2> Future Work </h2></summary>

Provide a description of any possible follow-up work that is explicitly
not being done in this pull request.
</details>
-->

<!--
<details>
<summary><h2> Bonus </h2></summary>

Provide a description of any extra problems you have solved in this pull
request.
</details>
-->

abhi2196 commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -79,9 +79,12 @@
 # satisfy external auditing of CIS security benchmarks.
```
We can update the comment to include sshd_config file update as well which is required for CIS benchmarks.

abhi2196 reviewed (commented):
Can we update the testing done section with the testing done for these changes?

- Manual checks to see if the changes are reflected in the engine, SSH is working fine for the `delphix` user etc.
- Blackbox `pre-checkin` suite to validate that tests are working fine.

sebroy commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,6 +250,14 @@
     - { key: "X11Forwarding", value: "no" }
     - { key: "HostKeyAlgorithms", value: "-ssh-rsa*" }
+    - { key: "MaxAuthTries", value: "6" }
+    - { key: "PermitRootLogin", value: "no"}
```
We've had the following setting for internal development engines for a while:
https://github.com/delphix/appliance-build/blob/a1b7edf79189d0718907826518c3b21450b9bfa7/live-build/misc/ansible-roles/appliance-build.virtualization-development/tasks/main.yml#L105
I don't think we need root logins at all, even for internal development engines, so it would be good to delete that directive from virtualization-development role as well.

justsanjeev replied:
We had to provide PermitRootLogin, the CIS test was failed without this.

sebroy replied:
Indeed, I had misread that as "yes", but it was "no", which is correct. https://delphix.atlassian.net/browse/DLPX-86536 does indeed include the requirement for `PermitRootLogin`.

sebroy commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,6 +250,14 @@
+    - { key: "MaxAuthTries", value: "6" }
+    - { key: "PermitRootLogin", value: "no"}
+    - { key: "Ciphers", value: "chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com"}
+    - { key: "MACs", value: "umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512"}
```
This conflicts with a different setting for `MACs` earlier in this file. If we're in agreement that we should use an explicit list of MACs (and not an exclusionary list like before), then the other code should be deleted.

rasantel commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,6 +251,12 @@
     - { key: "X11Forwarding", value: "no" }
     - { key: "HostKeyAlgorithms", value: "-ssh-rsa*" }
+    - { key: "MaxAuthTries", value: "6" }
+    - { key: "Ciphers", value: "chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com"}
```
This is already the default according to `man ssh_config`. I think it's safer to let sshd use whatever the safe defaults are at the time of each sshd release. Does CIS really complain if we don't set `Ciphers`? Maybe you can double check with Chris Price and hist team if it's ok with them to ignore CIS here and let sshd choose the default list.

sebroy replied:
@rasantel Yes, CIS does require that the settings be explicit even if they're equal to the defaults. The rationale is that the defaults could erroneously start including insecure settings. It does put the burden on us to update this list with new settings as they become available.

justsanjeev replied:
@rasantel Can I consider it resolved? feel free to reopen if required.

rasantel commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,6 +251,12 @@
     - { key: "HostKeyAlgorithms", value: "-ssh-rsa*" }
+    - { key: "MaxAuthTries", value: "6" }
+    - { key: "Ciphers", value: "chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com"}
+    - { key: "KexAlgorithms", value: "curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256"}
```
Same comment as with `Ciphers` above

sebroy commented:
@justsanjeev what happened to the MACs setting?

sebroy commented:
On the testing side:
1. Please include a link to a passing ab-pre-push run (this include blackbox pre-checkin which should also address Abhishek's comment)
2. Please run the CIS benchmark on an updated engine and show that the set of findings outlined in the bug are no longer issues.

ShibasishDelphix commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -266,23 +278,17 @@
-    # need to run long running processes over SSH on our internal
-    # systems (e.g. for development, testing, etc).
-    #
-    - variant is regex("external-.*")
```
As this is not linked to CRA, moving this check of external image looks fine.

justsanjeev replied:
No action expected, resolving.

ShibasishDelphix commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,17 +251,25 @@
 
 #
-# The CRA project mandated a 30 minute timeout for any idle connections.
+# The timeout is 5 minutes for any idle connections for AWS.
```
[nit] - Lets mentioned `timeout for client alive message' ? Hopefully that will help people not mix this up with connectionalive

sebroy commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -247,17 +251,23 @@
     - { key: "AllowTcpForwarding", value: "no" }
     - { key: "X11Forwarding", value: "no" }
     - { key: "HostKeyAlgorithms", value: "-ssh-rsa*" }
+    - { key: "MaxAuthTries", value: "4" }
```
Can we consolidate the different places in this file where we update sshd_config? There's this block here, and there's another one down on line 283 (following the `set_fact` settings). It feels like we should set the facts we nee to set, and then have a single task to update sshd_config?

justsanjeev replied:
Make sense, will do it!

justsanjeev replied:
Hi @sebroy ,

I have committed the changes suggested and have arranged the `keys` in alphabetical order as well.

`git ab-pre-push` - Done - http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/7825/

Once I get the build I can re-perform the Upgrade test from 19.0.0.0 to the upgrade artifact from the latest build, but I think that's not mandatory(in my opinion).

sebroy commented on files/common/var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml:
```diff
@@ -266,23 +276,17 @@
 - set_fact:
     ssh_client_alive_interval: "180"
-  when: platform == "azure"
+    ssh_client_alive_count_max: "0"
```
Why is this set to 0 for Azure?

justsanjeev replied:
Below are the current settings on 

Azure External Image
````
delphix@dqa19000-ext-azure:~$ sudo sshd -T | grep -i alive
clientaliveinterval 180
clientalivecountmax 0
tcpkeepalive yes
````

Azure Internal Image
````
delphix@dqa19000-azure:~$ sudo sshd -T | grep -i alive
clientaliveinterval 0
clientalivecountmax 3
tcpkeepalive yes
````


I tried to preserve a few settings, here although the `clientaliveinterval ` on Azure is recommended to `180`, however, max allowed is 300 which is set on `AWS`.

With some misunderstanding, the `clientaliveinterval` was set to `18000` on AWS internal image.

With this, change tried to keep the value standardized on Internal & External Images along with a few native settings.

Well, I am still open to further change and for discussions.

sebroy commented:
Can you also test upgrade? When you upgrade to an image with these changes, does the `sshd_config` file look exactly the same as on a newly installed system?

justsanjeev commented:
> Can you also test upgrade? When you upgrade to an image with these changes, does the `sshd_config` file look exactly the same as on a newly installed system?

It makes sense @sebroy , I will do it and note it for the TODO list for other CIS items.

justsanjeev commented:
`git ab-pre-push` - http://selfservice.jenkins.delphix.com/job/blackbox-chained/5861/
`Upgrade` - http://selfservice.jenkins.delphix.com/job/blackbox-chained/5872/

After the Upgrade, another round of scans is completed on the upgraded Engine. 
[Compliance_Report_new_report_10_110_248_191_dephx3sr1_20240122.pdf](https://github.com/delphix/delphix-platform/files/14005369/Compliance_Report_new_report_10_110_248_191_dephx3sr1_20240122.pdf)

justsanjeev commented:
- Latest `git ab-pre-push` - http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/7795/ - SUCCESSFUL 

- Upgrade Artifact path - http://selfservice.jenkins.delphix.com/job/appliance-build-stage0/job/pre-push/6827/consoleFull

- Used below to upgrade , from `18.0.0.0` to the below 

````
 aws s3 ls s3://dev-de-images/builds/jenkins-selfservice/appliance-build/develop/pre-push/2126/upgrade-artifacts/
2024-01-29 09:07:35    181 MD5SUMS
2024-01-29 09:07:35    277 SHA256SUMS
2024-01-29 09:07:35 5223772160 external-standard.upgrade.tar
2024-01-29 09:07:35 5415383040 internal-dev.upgrade.tar
2024-01-29 09:07:35 5285488640 internal-qa.upgrade.tar
````

Upgrade Chained Template -

````
---
settings:
  reuse_run_config: ORACLE_19.11.0.0.0_RHEL_8.4_SI
  num_servers: 1
steps:
  - version: 19.0.0.0
    group: dlpx-qa-19.0.0.0
    scopes:
    - run_type: setup
      terminate_pipeline_at_error: True
    - run_type: pre_upgrade_platform_modifier_tests
      extra_params: "-v"
    - run_type: pre_upgrade_platform_recorder_tests
      extra_params: "-v"
    - run_type: upgrade_only
      extra_params: "-p upgrade-target=20.0.0.0 -p custom-upgrade-image=s3://dev-de-images/builds/jenkins-selfservice/appliance-build/develop/pre-push/2126/upgrade-artifacts/internal-qa.upgrade.tar"
      terminate_pipeline_at_error: True
  - branch: develop
    qa_branch: develop
    scopes:
    - run_type: post_upgrade_platform_verifier_tests
      extra_params: "-v"
    - run_type: post_upgrade_platform_modifier_tests
      terminate_pipeline_at_error: True
```` 

Upgrade Result - http://selfservice.jenkins.delphix.com/job/blackbox-chained/5907/consoleFull - SUCCESS


Scan on the Upgraded Engine - All sections passed
[Compliance_Report_CIS_SCAN_10_110_202_79_dephx3sr1_20240130.pdf](https://github.com/delphix/delphix-platform/files/14095842/Compliance_Report_CIS_SCAN_10_110_202_79_dephx3sr1_20240130.pdf)

justsanjeev commented:
Created a `dlpx-18.0.0.0` engine, performed the steps below (as suggested)

````
Deploy a fresh VM without your changes on it (e.g. plain dlpx-develop).
Copy you version of main.yml to /var/lib/delphix-platform/ansible/10-delphix-platform/roles/delphix-platform/tasks/main.yml
Remove the /var/lib/delphix-platform/ansible-done file.
Restart the delphix-platform systemd service like this: systemctl restart delphix-platform
Check the contents of /etc/ssh/sshd_config and make sure it looks the same as a VM that was created from your bits (e.g. from the image generated by ab-pre-push).
````
Attached are the sshd_config contents - 
[dlpx-18000-sshd_config.txt](https://github.com/delphix/delphix-platform/files/14096288/dlpx-18000-sshd_config.txt)



Then, took the contents of the upgraded VM(http://dlpx-qa-19000-bb-chain-5907-8532dd44.dlpxdc.co/) from the previous step, here are the contents - 
[dlpx-19000-20000-upgrade-sshd_config.txt](https://github.com/delphix/delphix-platform/files/14096298/dlpx-19000-20000-upgrade-sshd_config.txt)

The files are identical(attached screenshot)  - 
<img width="1760" alt="Screenshot 2024-01-30 at 3 50 11 PM" src="https://github.com/delphix/delphix-platform/assets/82951964/087649ef-08b7-4cd9-8d06-9184bcd4744b">

Looks like we are good with the change, @sebroy , request you to have a look (last 2 comments) !

justsanjeev committed:
As per Seb's recomendations consolidated the settings for sshd
DLPX-86536 CIS: sshd configuration
Incorporating Shib's comment

PR URL: https://www.github.com/delphix/delphix-platform/pull/468

justsanjeev commented:
Upgrade from `19.0.0.0` to the upgrade artifact of the latest build http://selfservice.jenkins.delphix.com/job/blackbox-chained/5920/
is successful
 http://selfservice.jenkins.delphix.com/job/blackbox-chained/5920/consoleFull
